mat64/dense_arithmetic.go

```diff
@@ -407,11 +415,14 @@ func (m *Dense) Mul(a, b Matrix) {
 			w.mat.Data[r*w.mat.Stride+c] = v
 		}
 	}
-	*m = w
+	if w != m {
+		m.Copy(w)
+	}
 }
 
 // MulTrans takes the matrix product of a and b, optionally transposing each,
-// and placing the result in the receiver.
+// and placing the result in the receiver. If a or b are also the receiver, a
+// workspace will be allocated.
 //
 // See the MulTranser interface for more information.
 func (m *Dense) MulTrans(a Matrix, aTrans bool, b Matrix, bTrans bool) {
```

Comment:
Same here, I think instead we should check that m has the right size if it's non-zero.